subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock mtrace_lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_MTRACE_OUTPUT_DEFAULT;
+
+static adsp_mtrace_log_hook_t mtrace_hook;
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
+ *	------------------------
+ *	| Page1  - slot0       |
+ *	------------------------
+ *	| Page2  - slot1       |
+ *	------------------------
+ *	|         ...          |
+ *	------------------------
+ *	| Page14  - slot13     |
+ *	------------------------
+ *	| Page15  - slot14     |
+ *	------------------------
+ *
+ * The slot size == page size
+ *
+ * The first page contains descriptors for the remaining 15 cores
+ * The slot descriptor is:
+ * u32 res_id;
+ * u32 type;
+ * u32 vma;
+ *
+ * Log buffer slots have the following layout:
+ * u32 host_read_ptr;
+ * u32 dsp_write_ptr;
+ * u8 buffer[];
+ *
+ * The two pointers are offsets within the buffer.
+ */
+
+#define MTRACE_DESCRIPTOR_SIZE		12 /* 3 x u32 */
+#define MTRACE_PAGE_SIZE		0x1000
+#define MTRACE_SLOT_SIZE		MTRACE_PAGE_SIZE
+
+/* debug log slot types */
+#define MTRACE_SLOT_UNUSED			0x00000000
+#define MTRACE_SLOT_CRITICAL_LOG		0x54524300 /* byte 0: core ID */
+#define MTRACE_SLOT_DEBUG_LOG		0x474f4c00 /* byte 0: core ID */
+#define MTRACE_SLOT_GDB_STUB		0x42444700
+#define MTRACE_SLOT_TELEMETRY		0x4c455400
+#define MTRACE_SLOT_BROKEN			0x44414544
+ /* for debug and critical types */
+#define MTRACE_SLOT_CORE_MASK		GENMASK(7, 0)
+#define MTRACE_SLOT_TYPE_MASK		GENMASK(31, 8)
+
+#define MTRACE_LOG_BUF_SIZE		(MTRACE_SLOT_SIZE - 2 * sizeof(uint32_t))
+
+#define MTRACE_LOGGING_SLOT_TYPE(n)	(MTRACE_SLOT_DEBUG_LOG | (n & MTRACE_SLOT_CORE_MASK))
+
+#define MTRACE_DESCRIPTOR(n) ((uint8_t *)(z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE) + n * MTRACE_DESCRIPTOR_SIZE))
+
+#define MTRACE_LOG_BUF(n) (((uint8_t *)z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE)) + (n + 1) * MTRACE_SLOT_SIZE)
+
+static mem_addr_t mtrace_desc_type(int n)
+{
+	uint8_t *base = z_soc_uncached_ptr((void *)HP_SRAM_WIN2_BASE);
+	uint8_t *desc_base = base + n * MTRACE_DESCRIPTOR_SIZE;
+
+	return (mem_addr_t)(desc_base + 4);
+}
+
+static void mtrace_init(void)
+{
+        uint32_t desc_type_slot0 = sys_read32(mtrace_desc_type(0));
+
+	if (desc_type_slot0 == MTRACE_LOGGING_SLOT_TYPE(0))
+		return;
+
+	sys_write32(MTRACE_LOGGING_SLOT_TYPE(0), mtrace_desc_type(0));
+}
+
+static void mtrace_out(int8_t *str, size_t len, size_t *space_left)
+{
+	uint8_t *base = MTRACE_LOG_BUF(0);
+	uint8_t *data = base + 8;
+	uint32_t r = sys_read32((mem_addr_t)base);
+	uint32_t w = sys_read32((mem_addr_t)(base + 4));
```

Comment:
It's really just style, but I don't see value in sys_read() here.  Those are just SRAM words, not MMIO or special memory in any way, and those access macros might look like they do special cache management or atomicity or barrier work, but they don't.  They're just front ends for integer pointer dereferences, so I guess my eyes would want to see them expressed that way.

Reply:
Yeah, I actually had plain dereferences at first, but then converted to sys_read32(). sys_read32() provide me the "volatile" property, but yeah. maybe this is abuse (or at least a bit misleading). The  memory is a bit special as it is also accessed by another CPU, running on a different OS. I guess it doesn't matter here, but something like sys_read32() seems like a good way to remind readers that this is not normal DSP memory accesses. But let me check if I can align this more with surrounding Zephyr code.